### DISCUSSION TITLE
FRAGMOS asking : "what is exact meaning of annotation [calculation] ?"

### DISCUSSION THREAD

**Discussion by JBZ-Fragmos** [finos/common-domain-model#3425] [Q&A] [answered]
what is the exact meaning of this annotation `[calculation]` ? 
![image](https://github.com/user-attachments/assets/50b0565a-0509-40eb-8af4-f684ae91b879)

Documentation below says it means calculation is **ready-made** : 
![image](https://github.com/user-attachments/assets/82d186cb-0181-4330-897b-3cd03bc90f87)

 
However **do not understand** how it is exactly working, notably in regards of these two main sets of concerns : 

let' stay with example of `[calculation]` for `FloatingAmount`

- **question 1 :** Given the various features of rate returns payout terms (e.g. averaging, compounding, 2021 overnights styles, discounting, etc.) plus possible combinations of these terms… 
    - is the whole scope **exhaustively implemented** 
    - and **successfully tested**
    - and if yes, based on what **test file distribution** ?
    - and **test documentation** ?

- **question 2:** after first quick look I had across functional objects involved/composed together for eventually calculating the `FloatingAmount` and I could see lot of input are **“_external input_”** thus **_not being resolved from the Payout path values_** per se… that is to say there is no automation processing here because an extern
    - meaning there is actually **_no automation processing_**
    because there are **gaps** to manually fulfill, whereas should be resolving per functions, between **Event model** where agreements on terms is represented and related **Process model** based on “func” objects for the purpose of automating the execution of the Event model… as far as I can understand at this stage… if so, how to assess exactly where we stand in regards of target where ideally would have fully smart implementation i.e. “Trade -> Fixing -> “func” -> Calculation results” ?

**Comment by brianlynn2** [comment, marked as answer]
JB - In this, the [calculation] tag means that a full implementation of the specific function (but not of its supporting functions) is provided in Rosetta code, so that you don't need to override it with a Java or other language implementation.  We (@eacunaISDA  and I) did develop a sample implementation of the full calculations, including retrieval of sample supporting data (e.g. floating rate index observations, and business center holidays) for a demonstration application, which I believe remains available somewhere (but haven't looked recently).   This logic was also subsequently taken over by some of the folks at Trade Header for some additional experimentation, and I believe they've done demos of that at some industry events.

Doing this type of implementation is complex and raises many questions.  For example, what is the order of precedence of some of the calculations?    When we attempted to get answers to these questions at ISDA, none were forthcoming.  Basically it seems that no one knows for all of the many cases.  (E.g. what happens when have a multiplier and a spread and a cap on the same trade, which is also averaging and compounding? While there may be implementations out in the industry that cover that, the correct handling ISDA documentation is often unclear, because these combinations are relatively rare and apparently no one has come to ISDA for guidance.)  So the implementation is our best guess and the industry will need to confirm or correct.   I believe the release notes at the time explained this.

We did this sample implementation of the new (at the time in 2021) compounding logic for the industry to take a look at and provide feedback/criticize/refine.  We didn't have the resources to do all of the things you talked about.  Just to put a stake in the ground.  We hoped that subsequent implementers would be able to provide more complete solutions and fix any bugs/issues/limitations in what we developed.   (And doing even this little step was quite difficult due to limitations of the Rosetta syntax at the time, many of which we had to work around, and which were subsequently addressed.)

We did hope to provide better support for these types of calculations with some work at ISDA on reference data APIs to support CDM, but that effort was cancelled subsequent to the development of the sample calculation, for cost reasons.  So a lot of the things you were looking for never got funded, or had funding dropped.

So the short answer to your question is that no, many of those things haven't been done except in a demonstration application.  Turning it into a real, industrial-strength implementation will require much more resources than have been expended on this to date.

The [calculation] tag isn't a commitment that there is a fully working system, just a note that there is a Rosetta implementation of the calculation in the function.

Cheers,
Brian

**Comment by JBZ-Fragmos** [comment]
thanks a lot for clarifying @brianlynn2 

your feedback is aligned with my understanding about current business needs in that matter, 

based on which i'm currently working on several proposals, complementary to each other, which all together should hopefully improve the effectiveness and the reliability of the process model - just as an indication at this stage, assuming i cannot further elaborate here as part of this particular GitHub Issues, see below 3 proposals out of my list, and feel free to get i touch with me about it if that is raising interest for you : 

- the particular set of component i refer to as **"BespokePayoutFormula"** will notably permit to represent in Product model the exact calculation terms at stake in given payout terms https://github.com/finos/common-domain-model/issues/2941

  accordingly, a Trade for which BespokePayoutFormula has been populated will contain the formula to answer your  question above _"what is the order of precedence of some of the calculations? When we attempted to get answers to these questions at ISDA, none were forthcoming. Basically it seems that no one knows for all of the many cases. (E.g. what happens when have a multiplier and a spread and a cap on the same trade, which is also averaging and compounding? While there may be implementations out in the industry that cover that, the correct handling ISDA documentation is often unclear, because these combinations are relatively rare and apparently no one has come to ISDA for guidance"_ : **direct reading of the Payout path values in BespokePayoutFormula should provide the expected answers so you can draft an effective Process model based on clarified basis per Trade agreement at stake**

- another proposal i have is to **split current annotation [calculation]** into more specific ones, also to associated them with some **specific qualifiers** and some **specific references** in order to clarify scope of applicability of Product features with regards to scope of utilities functions in Process model... have not yet started to further document all this but in one word that is also a direction we may work on for improving Process model

- this one is a proposal for refactoring current set of attributes in Product model that is implicitly representing main overnight features with regards to 2021 ISDA Definition, that are at stake in understanding exact calculations shall be performed, etc. 
https://github.com/finos/common-domain-model/issues/3177
the rationale of the change proposal is to consider that 
   - there is a need for **clarification** : current attributes are quite far from the "drafting logic" of the 2021 Def which does not help to clarify what has to be calculated
   -  also because we found some **features are missing** : some terms from ISDA Def have not been digitized in the Product model so there are some gaps i.e. Trade for which you cannot simply represent exactly the overnight features, hence having no chance to apply any calculation if the source inputs does not even exist in the Product model where to represent the Trade agreement
